### ISSUE TITLE
Wrongly decoding uploaded filenames

### ISSUE THREAD

**Issue by Safihre** [cherrypy/cherrypy#1766]
**I'm submitting a ...**
- [x] bug report
- [ ] feature request
- [ ] question about the decisions made in the repository

**Do you want to request a *feature* or report a *bug*?**
Finally, we at SABnzbd are converting to Python 3!
I thought this would solve an issue that we had for a long time, that I attributed to Python 2.7:
Maybe it is the same problem as @ngrilly in #1726, but when I upload [你好安全.txt](https://github.com/cherrypy/cherrypy/files/2796942/default.txt) to CherryPy running on that same machine, it is wrongly decoded to have a `filname` of `ä½ å¥½å®å¨`.

**What is the current behavior?**
It seems to use the wrong decoding.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a screenshots and logs of the problem. If you can, show us your code.**
Upload the text-file linked above.

**What is the expected behavior?**
To see the correct filename. This seems to be possible, judging from also GitHub here correctly detecting the filename.

**What is the motivation / use case for changing the behavior?**
I want to see the correct filename.

**Please tell us about your environment:**

- Cheroot version: 6.5.2
- CherryPy version: 18.1.0
- Python version: 3.7.2
- OS: Windows 10 and Ubuntu
- Browser: all

**Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, e.g. stackoverflow, gitter, etc.)

**Comment by webknjaz** [comment]
Is it Windows-specific? Could you please send a PR with a test triggering this?

**Comment by webknjaz** [comment]
Possible reproducer: https://github.com/cherrypy/cherrypy/pull/1434/files#diff-d8e96680529ae6d52d1183209f09bef1R301

**Comment by Safihre** [comment]
You are right, it's not just Windows I can also reproduce it on my Ubuntu 18.04 VM with latest available Firefox and Chrome.

**Comment by Safihre** [comment]
Not sure how to reproduce in a pretty test. I checked on Windows and Ubuntu, both browsers show:
```
Content-Disposition: form-data; name="name"; filename="你好安全.nzb"
```
Not sure how this then translates in the request that CherryPy processes.

But with this info, you should be able to easily reproduce locally?

**Comment by Safihre** [comment]
I can confirm that https://github.com/cherrypy/cherrypy/pull/1434/files#diff-d8e96680529ae6d52d1183209f09bef1R301 is the reproducer.
The filename in this issue would be `\u4f60\u597d\u4e16\u754c`.

**Comment by Safihre** [comment]
@webknjaz Any thoughts on this? I wouldn't mind contributing, but I'm a bit of a loss how the whole system comes together.

**Comment by pawciobiel** [comment]
I have the same problem on both python 27 cherrypy 17.4.1 and python3.7 cherrypy 18.1.0
It looks like filenames are not properly decoded into unicode (str in py3) object.
Fields from multipart/form-data are decoded correctly.
```
(Pdb++) myFile.filename
u'B\xc5\x82o\xc5\x84skiego.png'
(Pdb++) dupa
u'\u017c\u017a\u0107\u0142\xf3\u0119\u0105'
```

```
# coding: utf-8
class FileDemo(object):

    @cherrypy.expose
    def index(self):
        return """
        <html>
        <head>
        <meta charset="UTF-8">
        </head>
        <body>
            <h2>Upload a file</h2>
            <form action="upload" method="post" enctype="multipart/form-data" accept-charset="UTF-8">
            filename: <input type="file" name="myFile" /><br />
            <input type="text" name="dupa" value="żźćłóęą"/><br />
            <input type="submit" />
            </form>
        </body></html>

        """

    @cherrypy.expose
    def upload(self, myFile=None, dupa=None):
        logging.error(type(myFile.filename))
        logging.error(myFile.filename)
        logging.error(type(dupa))
        logging.error(dupa)
```

**Comment by pawciobiel** [comment]
@Safihre you can workaround the problem by using client which will use Content-Disposition percent-encoding and character encoding:
```
Content-Disposition: attachment; filename*=UTF-8''Na%C3%AFve%20file.txt
```

Sadly this won't work for basic multipart form file upload.

**Comment by Safihre** [comment]
It's been a year, any more thoughts on this @webknjaz ?

**Comment by webknjaz** [comment]
Ah, sorry. Lost track of it. Feel free to contribute. I'd start with extracting that test function and sending it as a PR but mark it with xfail as per https://pganssle-talks.github.io/xfail-lightning. This way, we'd have something actually running in CI. After that, it'd be easier to start fixing it (remove that xfail marker and ensure it's always green).

You'd need to verify what's in the actual payload (what browsers show in the UI may be some recoded repr, not the real bytes). Also, we need some RFC pointers confirming that it's actually allowed to send raw bytes like that.

Try to make the patch non-intrusive and make sure it plays nice with `tools.encode`/`tools.decode`. Their source is https://github.com/cherrypy/cherrypy/blob/242dc04/cherrypy/lib/encoding.py#L12-L256 and they are exposed via https://github.com/cherrypy/cherrypy/blob/242dc04/cherrypy/_cptools.py#L479-L481.
